Add `Scarb.toml` docs and template generation (#2516)
foundry-rs/starknet-foundry: docs/src/appendix/scarb-toml.md

# The Manifest Format

> The `Scarb.toml` contains the package manifest that is needed in package compilation process. It can be used to provide configuration for Starknet Foundry Forge.

Reviewer: Maybe let's also link to relevant scarb doc page here?

## `Scarb.toml` Contents

### `[tool.snforge]`
```toml
[tool.snforge]
# ...
```
Allows to configure `snforge` settings. All fields are optional.

#### `exit_first`
The `exit_first` fields specifies whether to stop tests execution immediately upon the first failure.
```toml
[tool.snforge]
exit_first = true
```

#### `fuzzer_runs`
The `fuzzer_runs` field specifies the number of runs of the random fuzzer.

#### `fuzzer_seed`
The `fuzzer_seed` field specifies the seed for the random fuzzer.

#### Example of fuzzer configuration
```toml
[tool.snforge]
fuzzer_runs = 1234

> fuzzer_seed = 1111

Reviewer: Let's link to fuzzer section in docs

```

### `[[tool.snforge.fork]]`
```toml
[[tool.snforge.fork]]
# ...
```

> Allows to configure forked tests. While in use, all below fields must be set at the same time.

Reviewer: ```suggestion
Allows to configure forked tests. If defined, all fields outlined below must also be defined.
```

#### `name`
The `name` field specifies the name of the fork.
```toml
[[tool.snforge.fork]]
name = "SOME_NAME"
```

#### `url`
The `url` field specifies the address of RPC provider.
```toml
[[tool.snforge.fork]]
url = "http://your.rpc.url"
```

#### `block_id.<tag|number|hash>`
The `block_id` field specifies the block to fork from. It can be specified by `tag`, `number` or `hash`.
```toml
[[tool.snforge.fork]]
block_id.hash = "0x123"
```

#### Example configuration with two forks
```toml
[[tool.snforge.fork]]
name = "SOME_NAME"
url = "http://your.rpc.url"
block_id.tag = "latest"

> 

Reviewer: Let's link fork docs

[[tool.snforge.fork]]
name = "SOME_SECOND_NAME"
url = "http://your.second.rpc.url"
block_id.number = "123"
```

### `[profile.<dev|release>.cairo]`

> Adjust Cairo compiler configuration parameters when compiling this package. These options are not taken into consideration when this package is used as a dependency for another package. All fields are optional.

Reviewer: ```suggestion
By default, these arguments do not need to be defined. Only set them to use profiler (link docs) or coverage (link docs).

Adjust Cairo compiler configuration parameters when compiling this package. These options are not taken into consideration when this package is used as a dependency for another package. All fields are optional.
```

```toml
[profile.dev.cairo]
unstable-add-statements-code-locations-debug-info = true
```

#### `unstable-add-statements-functions-debug-info`
If enabled, during the project compilation Scarb will a add mapping between Sierra statement indexes and vectors of fully qualified paths of Cairo functions to debug info. A statement index maps to a vector consisting of a function which caused the statement to be generated and all functions that were inlined or generated along the way. By default, this flag is disabled.
```toml
[profile.dev.cairo]
unstable-add-statements-functions-debug-info = true
```

#### `inlining-strategy`
This field is responsible for setting the inlining strategy to be used by compiler when building the package. The possible values are `default` or `avoid`. If avoid strategy is set, the compiler will only inline function annotated with `#[inline(always)]` attribute.
```toml
[profile.dev.cairo]
inlining-strategy = "avoid"

> ```

Reviewer: Maybe let's just link to specific sections of scarb docs here instead of maintaining these ourselves?

Scarb supports registering targets that are handled by Scarb extensions. Such targets are called external. All fields are optional.
```toml
[[target.starknet-contract]]
# ...
```

#### `sierra`
The `sierra` fields specifies whether Sierra codegen should be enabled.
```toml
[[target.starknet-contract]]
sierra = true
```

#### `casm`
The `casm` fields specifies whether Casm codegen should be enabled.
```toml
[[target.starknet-contract]]
casm = true
```

#### `build-external-contracts`
The `build-external-contracts` accepts a list of strings, each of which is a reference to a contract defined in a dependency. The package that implements this contracts need to be declared as a dependency of the project in `[dependencies]`.
```toml
[[target.starknet-contract]]
build-external-contracts = ["openzeppelin::account::account::Account"]
```

#### Example of configuration which allows to use external contracts in tests
```toml
# ...
[dependencies]
starknet = ">=2.8.2"
openzeppelin = { git = "https://github.com/OpenZeppelin/cairo-contracts.git", branch = "cairo-2" }

[[target.starknet-contract]]
build-external-contracts = ["openzeppelin::account::account::Account"]
# ...

> ```

Reviewer: While I get why this was included, this is kind of duplicating Scarb docs and adds an additional burden for us of maintaining it. 

Maybe let's just say that we handle `[features] and link to both a section in our docs that explain conditional compilation and to scarb docs? Similar on how scarb does it in their reference

<img width="831" alt="image" src="https://github.com/user-attachments/assets/107ac8d9-2e50-4616-8c64-d75983e5a71a">

Same goes for starknet contract target and maybe even for `unstable-...`.

```toml
[profile.dev.cairo]
inlining-strategy = "avoid"
```

#### Example of configuration which allows [coverage](https://foundry-rs.github.io/starknet-foundry/testing/coverage.html) report generation
```toml
[profile.dev.cairo]
unstable-add-statements-code-locations-debug-info = true
unstable-add-statements-functions-debug-info = true
inlining-strategy = "avoid"
```

### `[features]`
A package defines a set of named features in the `[features]` section of `Scarb.toml` file. Each defined feature can list other features that should be enabled with it. All fields are optional.
```toml
[features]
# ...
```

#### `<feature-name>`
The `<feature-name>` field specifies the name of the feature and list of other features that should be enabled with it.
```toml
[features]
enable_for_tests = []
```

#### Example of `Scarb.toml` allowing conditional contracts compilation
Firstly, define a contract in the src directory with a `#[cfg(feature: '<FEATURE_NAME>')]` attribute:
```rust
#[starknet::contract]
#[cfg(feature: 'enable_for_tests')]
mod MockContract {
    // ...
}
```

Then update Scarb.toml so it includes the following lines:
```toml
[features]
enable_for_tests = []
```

> See [features](https://docs.swmansion.com/scarb/docs/reference/conditional-compilation.html#features) in Scarb docs.

Reviewer: Let's also mention this at the top of the section

### `[[target.starknet-contract]]`
Scarb supports registering targets that are handled by Scarb extensions. Such targets are called external. All fields are optional.
```toml
[[target.starknet-contract]]
# ...
```

#### `sierra`
The `sierra` fields specifies whether Sierra codegen should be enabled.
```toml
[[target.starknet-contract]]
sierra = true
```

#### `casm`
The `casm` fields specifies whether Casm codegen should be enabled.
```toml
[[target.starknet-contract]]
casm = true
```

See more about [targets](https://docs.swmansion.com/scarb/docs/reference/targets.html) in scarb docs.

#### `build-external-contracts`

> The `build-external-contracts` accepts a list of strings, each of which is a reference to a contract defined in a dependency. The package that implements this contracts need to be declared as a dependency of the project in `[dependencies]`.

Reviewer: This is a super formal description but doesn't really say what's the purpose of this in terms of forge.

Let's say that this is needed to use the contracts from dependencies in tests and also link to relevant scarb doc. You can keep the mention that these packages also need to be defined as dependencies, but this can be reworded.

Author: Hope this is better.

> The `build-external-contracts` allows to use contracts from your dependencies inside your tests. It accepts a list of strings, each of which is a reference to a contract defined in a dependency. You need to add dependency which  implements this contracts  to your `Scarb.toml`.

I've added a sentence which explains what this option actually does in regards of forge, should be clearer now.

A package defines a set of named features in the `[features]` section of `Scarb.toml` file. Each defined feature can list other features that should be enabled with it. All fields are optional.
```toml
[features]
# ...
```

#### `<feature-name>`
The `<feature-name>` field specifies the name of the feature and list of other features that should be enabled with it.
See [features](https://docs.swmansion.com/scarb/docs/reference/conditional-compilation.html#features) in Scarb documentation.
```toml
[features]
enable_for_tests = []
```

#### Example of `Scarb.toml` allowing conditional contracts compilation
Firstly, define a contract in the src directory with a `#[cfg(feature: '<FEATURE_NAME>')]` attribute:
```rust
#[starknet::contract]
#[cfg(feature: 'enable_for_tests')]
mod MockContract {
    // ...
}
```

Then update Scarb.toml so it includes the following lines:
```toml
[features]
enable_for_tests = []
```

### `[[target.starknet-contract]]`
Scarb supports registering targets that are handled by Scarb extensions. Such targets are called external. All fields are optional. See more about [targets](https://docs.swmansion.com/scarb/docs/reference/targets.html) in Scarb documentation.

> 

Reviewer: Let's just focus on `starknet-contract` target and say that this is required to build starknet contracts and link to this https://docs.swmansion.com/scarb/docs/extensions/starknet/contract-target.html#starknet-contract-target page. 

`sierra` and `casm` can just link to relevant subsections.

Author: Ok, looks as below now:
<img width="823" alt="image" src="https://github.com/user-attachments/assets/79d25282-cd18-49b7-a55e-05f3ded4ab85">
